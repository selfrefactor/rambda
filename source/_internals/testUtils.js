import combinate from 'combinate'
import {
  equals,
  filter,
  forEach,
  init,
  last,
  omit,
  switcher,
  type,
} from 'rambdax'

const omitOk = omit('ok')

const PENDING = 'PENDING'
const RESULTS_EQUAL = 'results are equal'
const RESULTS_MISMATCH = 'results are different'
const ERRORS_EQUAL = 'errors are equal'
const ERRORS_MESSAGE_MISMATCH = 'errors messages are different'
const ERRORS_TYPE_MISMATCH = 'errors types are different'
const SHOULD_THROW = 'Rambda should throw'
const SHOULD_NOT_THROW = 'Rambda should not throw'
const ALL_ERROR_LABELS = {
  RESULTS_MISMATCH,
  ERRORS_MESSAGE_MISMATCH,
  ERRORS_TYPE_MISMATCH,
  SHOULD_NOT_THROW,
  SHOULD_THROW,
}
const MISSING = 'Missing error handle in parseError'

function parseError(err){
  const typeError = switcher(err)
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof SyntaxError, 'SyntaxError')
    .is(x => x instanceof RangeError, 'RangeError')
    .is(x => x instanceof Error, 'Error')
    .default(MISSING)

  if (typeError === MISSING){
    throw new Error('typeError === MISSING')
  }

  return {
    message : err.message,
    type    : typeError,
    ok      : true,
  }
}

function executeSync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      fn(...init(inputs))(last(inputs)) :
      fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

async function executeAsync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      await fn(...init(inputs))(last(inputs)) :
      await fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

export function profileMethod({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const { result, error } = executeSync(
        fn, inputs, returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function profileMethodAsync({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)
    test(getTestTitle(...inputs), async () => {
      const { result, error } = await executeAsync(
        fn,
        inputs,
        returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function compareToRamda(
  fn, fnRamda, returnsFunctionFlag
){
  return (...inputs) => {
    const { result, error } = executeSync(
      fn, inputs, returnsFunctionFlag
    )
    const { result: ramdaResult, error: ramdaError } = executeSync(
      fnRamda,
      inputs,
      returnsFunctionFlag
    )

    const toReturn = {
      result,
      ramdaResult,
      ramdaError : ramdaError.ok ? omitOk(ramdaError) : PENDING,
      error      : error.ok ? omitOk(error) : PENDING,
    }

    if (result !== PENDING){
      if (ramdaError.ok){
        return {
          ...toReturn,
          ok    : false,
          label : SHOULD_THROW,
        }
      }

      if (equals(result, ramdaResult)){
        return {
          ...toReturn,
          ok    : true,
          label : RESULTS_EQUAL,
        }
      }

      return {
        ...toReturn,
        ok    : false,
        label : RESULTS_MISMATCH,
      }
    }

    if (equals(error, ramdaError)){
      return {
        ...toReturn,
        ok    : true,
        label : ERRORS_EQUAL,
      }
    }

    if (ramdaError.ok){
      return {
        ...toReturn,
        ok    : false,
        label :
          ramdaError.type === error.type ?
            ERRORS_MESSAGE_MISMATCH :
            ERRORS_TYPE_MISMATCH,
      }
    }

    return {
      ...toReturn,
      ok    : false,
      label : SHOULD_NOT_THROW,
    }
  }
}

export const getTestTitle = (...inputs) => inputs.map(type).join(' | ')

export const compareCombinations = ({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  setCounter = () => {},
  callback = x => {},
  fn,
  fnRamda,
}) => {
  const counter = {
    RESULTS_MISMATCH        : 0,
    SHOULD_THROW            : 0,
    SHOULD_NOT_THROW        : 0,
    ERRORS_TYPE_MISMATCH    : 0,
    ERRORS_MESSAGE_MISMATCH : 0,
    TOTAL_TESTS             : 0,
  }

  const increaseCounter = comparedResult => {
    if (comparedResult.ok) return
    let counterProp
    forEach((x, prop) => {
      if (x === comparedResult.label) counterProp = prop
    }, ALL_ERROR_LABELS)
    counter[ counterProp ]++
  }

  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)
  const compareOutputs = compareToRamda(
    fn, fnRamda, returnsFunctionFlag
  )

  afterAll(() => callback(counter))

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const compared = compareOutputs(...inputs)
      setCounter()
      if (!compared.ok){
        // if (compared.label === RESULTS_MISMATCH){
        increaseCounter(compared)
        expect({
          ...compared,
          inputs,
        }).toMatchSnapshot()
      }
      counter.TOTAL_TESTS++
    })
  })
}

const demoMap = new Map()

demoMap.set('a', 1)
demoMap.set('b', 2)

const demoWeakMap = new WeakMap()

demoWeakMap.set({ a : 0 }, 1)
demoWeakMap.set({ b : 0 }, 2)

const demoWeakSet = new WeakSet()
const foo = { a : 1 }
const bar = { b : 2 }

demoWeakSet.add(foo)
demoWeakSet.add(bar)

const demoBigInt = BigInt(9007199254740991)

const demoRegExp = new RegExp('ab+c', 'i')
const demoSymbol = Symbol('foo')

export const FALSY_VALUES = [ null, undefined, false, NaN, '' ]
export const EXTRA_BUILD_IN_OBJECTS = [
  Infinity,
  -Infinity,
  Symbol,
  new ArrayBuffer(8),
  demoRegExp,
  demoMap,
  demoWeakMap,
  demoWeakSet,
  demoBigInt,
  demoSymbol,
]
