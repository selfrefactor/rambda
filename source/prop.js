export function prop(key, obj){
  if (arguments.length === 1) return _obj => prop(key, _obj)

  if (!obj) return undefined

  return obj[ key ]
}
