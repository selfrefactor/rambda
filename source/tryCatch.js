import { isFunction } from './isFunction'
import { isPromise } from './isPromise'

export function tryCatch(fn, fallback){
  if (!isFunction(fn)){
    throw new Error(`R.tryCatch | fn '${ fn }'`)
  }
  const passFallback = isFunction(fallback)

  if (!isPromise(fn)){
    return (...inputs) => {
      try {
        return fn(...inputs)
      } catch (e){
        return passFallback ? fallback(...inputs) : fallback
      }
    }
  }

  return (...inputs) =>
    new Promise(resolve => {
      fn(...inputs)
        .then(resolve)
        .catch(() => {
          if (!passFallback){
            return resolve(fallback)
          }

          if (!isPromise(fallback)){
            return resolve(fallback(...inputs))
          }

          fallback(...inputs).then(resolve)
        })
    })
}
