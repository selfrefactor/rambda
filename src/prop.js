/**
 * Returns a function that when supplied an object returns the indicated
 * property of that object, if it exists.
 *
 * @func
 * @category Object
 * @sig s -> {s: a} -> a | Undefined
 * @param {String} key The property name
 * @param {Object} obj The object to query
 * @return {*} The value at `obj[key]`.
 * @example
 *
 *      R.prop('x', {x: 100}); //=> 100
 *      R.prop('x', {}); //=> undefined
 *      R.compose(R.inc, R.prop('x'))({ x: 3 }) //=> 4
 */
export function prop(key, obj){
  if (arguments.length === 1) return _obj => prop(key, _obj)

  if (!obj) return undefined

  return obj[ key ]
}
