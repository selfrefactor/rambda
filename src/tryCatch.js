import {isFunction} from './isFunction'

export function tryCatch(fn, fallback) {
  if (!isFunction(fn)) {
    throw new Error(`R.tryCatch | fn '${fn}'`)
  }
  const passFallback = isFunction(fallback)

  return (...inputs) => {
    try {
      return fn(...inputs)
    } catch (e) {
      return passFallback ? fallback(e, ...inputs) : fallback
    }
  }
}
