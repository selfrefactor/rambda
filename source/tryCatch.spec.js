import { tryCatch as tryCatchRamda } from 'ramda'

import { compareCombinations } from './_internals/testUtils.js'
import { prop } from './prop.js'
import { tryCatch } from './tryCatch.js'

test('happy', () => {
  const fn = () => {
    throw new Error('foo')
  }
  const result = tryCatch(fn, () => true)()
  expect(result).toBeTrue()
})

test('when fallback is used', () => {
  const fn = x => x.x

  expect(tryCatch(fn, false)(null)).toBeFalse()
})

test('with json parse', () => {
  const good = () => JSON.parse(JSON.stringify({ a : 1 }))
  const bad = () => JSON.parse('a{a')

  expect(tryCatch(good, 1)()).toEqual({ a : 1 })
  expect(tryCatch(bad, 1)()).toBe(1)
})

test('when fallback is function', () => {
  const fn = x => x.x

  expect(tryCatch(fn, () => 1)(null)).toBe(1)
})

test('when fn is used', () => {
  const fn = prop('x')

  expect(tryCatch(fn, false)({})).toBeUndefined()
  expect(tryCatch(fn, false)({ x : 1 })).toBe(1)
})

test('fallback receives error object and all initial inputs', () => {
  function thrower(
    a, b, c
  ){
    void c
    throw new Error('throwerError')
  }

  function catchFn(
    e, a, b, c
  ){
    return [ e.message, a, b, c ].join('|')
  }

  const willThrow = tryCatch(thrower, catchFn)
  const result = willThrow(
    'A', 'B', 'C'
  )
  expect(result).toBe('throwerError|A|B|C')
})

test('fallback receives error object', () => {
  function throwFn(){
    throw new Error(10)
  }

  function eCatcher(
    e, a, b
  ){
    return e.message
  }

  const willThrow = tryCatch(throwFn, eCatcher)
  expect(willThrow([])).toBe('10')
  expect(willThrow([ {}, {}, {} ])).toBe('10')
})

const possibleFns = [
  null,
  () => 1,
  () => 0,
  () => JSON.parse('{a:1'),
  () => {
    const x = {}

    return x.x
  },
  x => x.foo,
  () => {
    throw new Error('foo')
  },
]

const possibleCatchers = [
  null,
  e => e.message.length,
  (e, ...inputs) => `${ e.message.length } ${ inputs.length }`,
  () => {
    throw new Error('bar')
  },
]

const possibleInputs = [ null, {}, { foo : 1 } ]

describe('brute force', () => {
  compareCombinations({
    returnsFunctionFlag : true,
    firstInput          : possibleFns,
    callback            : errorsCounters => {
      expect(errorsCounters).toMatchInlineSnapshot(`
        {
          "ERRORS_MESSAGE_MISMATCH": 0,
          "ERRORS_TYPE_MISMATCH": 12,
          "RESULTS_MISMATCH": 0,
          "SHOULD_NOT_THROW": 0,
          "SHOULD_THROW": 7,
          "TOTAL_TESTS": 84,
        }
      `)
    },
    secondInput : possibleCatchers,
    thirdInput  : possibleInputs,
    fn          : tryCatch,
    fnRamda     : tryCatchRamda,
  })
})
