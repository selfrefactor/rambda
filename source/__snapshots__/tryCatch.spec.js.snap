// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brute force Function | Null | Null 1`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    null,
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Function | Null | Null 2`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    null,
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Function | Null | Null 3`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    null,
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Function | Null | Object 1`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    {},
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Function | Null | Object 2`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    {},
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Function | Null | Object 3`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    {
      "foo": 1,
    },
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Function | Null | Object 4`] = `
{
  "error": "PENDING",
  "inputs": [
    [Function],
    null,
    {
      "foo": 1,
    },
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'apply')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": null,
}
`;

exports[`brute force Null | Function | Null 1`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Null 2`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Null 3`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Object 1`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    {},
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Object 2`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    {},
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Object 3`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    {},
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Object 4`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    {
      "foo": 1,
    },
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Object 5`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    {
      "foo": 1,
    },
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Function | Object 6`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    [Function],
    {
      "foo": 1,
    },
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Null | Null 1`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    null,
    null,
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Null | Object 1`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    null,
    {},
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Null | Object 2`] = `
{
  "error": {
    "message": "R.tryCatch | fn 'null'",
    "type": "Error",
  },
  "inputs": [
    null,
    null,
    {
      "foo": 1,
    },
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'length')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;
